brokerbot.go

```diff
@@ -120,95 +122,89 @@ func initTokens() {
 }
 
 func handleDefaultPort(w http.ResponseWriter, r *http.Request) {
-	log.Println("Heartbeat")
-	timeSinceLastHeartbeat = time.Now()
-	fmt.Fprintln(w, "Hello World!")
+	fmt.Fprintln(w, "OK")
 }
 
 func handleMessage(s *discordgo.Session, m *discordgo.MessageCreate) {
-	log.Printf("time since last heartbeat: %s", time.Since(timeSinceLastHeartbeat))
-	/* Validation */
 	if m.Author.ID == s.State.User.ID {
-		// Prevent the bot from talking to itself.
+		// Ignore messages from self.
 		return
 	}
 
-	isBotMention, msg := hasBotMention(s, m)
+	splitMsg := strings.Fields(m.ContentWithMentionsReplaced())
+
+	if splitMsg[0] != botHandle && splitMsg[0] != botPrefix {
+		// Message wasn't meant for us.
+		return
+	}
 
-	if !isBotMention {
-		// Ignore bot mentions.
+	if len(splitMsg) < 2 || splitMsg[1] == helpToken {
+		// Message didn't have enough parameters.
+		messagelib.SendMessage(s, m.ChannelID, getHelpMessage())
 		return
 	}
 
-	msg = strings.TrimSpace(msg)
-	userInput := strings.TrimPrefix(msg, "!stonks ")
-	userInput = strings.ToUpper(userInput)
-	expandedString := messagelib.ExpandAliases(userInput)
-	tickers := messagelib.DedupeTickerStrings(strings.Split(expandedString, " "))
+	var tickers []string = splitMsg[1:]
+	tickers = messagelib.RemoveMentions(tickers)
+	tickers = messagelib.CanonicalizeMessage(tickers)
+	tickers = messagelib.ExpandAliases(tickers)
+	tickers = messagelib.DedupeSlice(tickers)
+
+	startTime := time.Now()
+	log.Printf("Received request for tickers: %s", tickers)
+
+	log.Println("Fetching crypto price feeds.")
 
 	priceFeeds, err := cryptolib.GetPriceFeeds(geminiClient)
 	if err != nil {
 		// We don't fail, because we gracefully handle no crypto price feeds in the
 		// handle crypto methods, and a request might contain multiple stock tickers
-		log.Printf("Failed to fetch Price Feeds: %s", err);
+		log.Printf("Failed to fetch crypto price feeds: %v", err)
 	}
 
-	if len(tickers) == 1 && tickers[0] == "" {
-		// TODO: Send a help message to the user.
-		log.Println("No stock tickers provided")
-		return
-	} else if len(tickers) == 1 && tickers[0] != "" {
-		log.Printf("Processing request for: %s", tickers[0])
-
-		tickerType, ticker := getTickerWithType(tickers[0])
+	var tickerValues []*messagelib.TickerValue
+	// TODO(matthewlavine): Make concurrent.
+	for _, rawTicker := range tickers {
+		ticker, tickerType := getTickerAndType(rawTicker)
 
 		switch tickerType {
 		case stock:
-			stocklib.HandleStockTicker(ctx, finnhubClient, s, m, ticker)
+			tickerValue, err := stocklib.GetQuoteForStockTicker(ctx, finnhubClient, ticker)
+			if err != nil {
+				msg := fmt.Sprintf("Failed to get quote for stock ticker: %q (See logs)", ticker)
+				log.Printf(fmt.Sprintf("%s: %v", msg, err))
+				messagelib.SendMessage(s, m.ChannelID, msg)
```

Comment:
So this will dispatch errors for every Stock and Crypto that isn't found? Can we also bundle these into a message Embed?

Reply:
These would be HTTP errors, not "not found" errors. Those are still handled transparently and returned as "no data" later on.

This error case has yet to be triggered in practice so so can change it if needed.